WebUI: Add new identity type (SAML) in core

Ruturaj123 opened the pull request:
WebUI: Add new identity type (SAML) in core

Ruturaj123 committed:
WebUI: Add new identity type (SAML) in core

bari12 commented:
Hi Ruturaj, there were some errors showing up in the travis test; Please fix 😄 

```
lib/rucio/web/ui/common/utils.py:96:57: E128 continuation line under-indented for visual indent
lib/rucio/web/ui/common/utils.py:97:57: E128 continuation line under-indented for visual indent
lib/rucio/web/ui/common/utils.py:98:57: E128 continuation line under-indented for visual indent
lib/rucio/web/ui/common/utils.py:128:57: E128 continuation line under-indented for visual indent
lib/rucio/web/ui/common/utils.py:129:57: E128 continuation line under-indented for visual indent
lib/rucio/web/ui/common/utils.py:130:57: E128 continuation line under-indented for visual indent
lib/rucio/api/authentication.py:150:1: E302 expected 2 blank lines, found 1
lib/rucio/core/authentication.py:310:1: W293 blank line contains whitespace
```

Ruturaj123 commented:
Yes Sir I have fixed the errors locally. Actually I have been getting some errors in the webui so I will push the changes once those errors have been fixed.

Ruturaj123 committed:
WebUI: Add SSO login option to WebUI

WebUI: Add SSO login option to WebUI

WebUI: Make python3-saml dependency optional

WebUI: Travis Fix

WebUI: Remove unwanted imports

Ruturaj123 committed:
WebUI: Add test cases

WebUI: Syntax fix

WebUI: Incorporate suggested changes

WebUI: Travis fix

mlassnig commented on bin/rucio:
```diff
@@ -241,7 +242,7 @@ def get_client(args):
     Returns a new client object.
     """
-    if args.auth_strategy == 'userpass':
+    if args.auth_strategy == 'userpass' or args.auth_strategy == 'saml':
```
`if args.auth_strategy in ['userpass', 'saml']:`

mlassnig commented on lib/rucio/client/baseclient.py:
```diff
@@ -152,8 +153,8 @@ def __init__(self, rucio_host=None, auth_host=None, account=None, ca_cert=None,
             if 'RUCIO_AUTH_TYPE' in environ:
-                if environ['RUCIO_AUTH_TYPE'] not in ('userpass', 'x509', 'x509_proxy', 'gss', 'ssh'):
-                    raise MissingClientParameter('Possible RUCIO_AUTH_TYPE values: userpass, x509, x509_proxy, gss, ssh, vs. ' + environ['RUCIO_AUTH_TYPE'])
+                if environ['RUCIO_AUTH_TYPE'] not in ('userpass', 'x509', 'x509_proxy', 'gss', 'ssh', 'saml'):
```
should be list, not tuple (original was wrong already)

mlassnig commented on lib/rucio/client/baseclient.py:
```diff
@@ -165,7 +166,7 @@ def __init__(self, rucio_host=None, auth_host=None, account=None, ca_cert=None,
             self.creds = {}
             try:
-                if self.auth_type == 'userpass':
+                if self.auth_type == 'userpass' or self.auth_type == 'saml':
```
`if self.auth_type in ['userpass', 'saml']:`

mlassnig commented on lib/rucio/client/baseclient.py:
```diff
@@ -538,6 +539,49 @@ def __get_token_gss(self):
+
+        headers = {'X-Rucio-Account': self.account}
+        userpass = {'username': self.creds['username'], 'password': self.creds['password']}
+        url = build_url(self.auth_host, path='auth/sso')
```
maybe the endpoint should actually be called `auth/saml` ? the mental jump from sso to saml is not obvious for people not used to AAI

mlassnig commented on lib/rucio/core/authentication.py:
```diff
@@ -301,6 +302,26 @@ def get_ssh_challenge_token(account, appid, ip=None, session=None):
+        return None
+
+    # remove expired tokens
+    session.query(models.Token).filter(models.Token.expired_at < datetime.datetime.utcnow(),
```
please add the `for_update` statement, like in https://github.com/rucio/rucio/pull/2721/files

mlassnig commented on lib/rucio/core/permission/atlas.py:
```diff
@@ -314,6 +315,19 @@ def perm_get_auth_token_x509(issuer, kwargs):
 
+def perm_get_auth_token_saml(issuer, kwargs):
+    """
+    Checks if a user can request a token with user_pass for an account.
```
mean to say saml here

mlassnig commented on lib/rucio/db/sqla/migrate_repo/versions/1c45d9730ca6_increase_identity_length.py:
```diff
@@ -81,13 +82,13 @@ def downgrade():
         create_check_constraint(constraint_name='IDENTITIES_TYPE_CHK',
                                 table_name='identities',
-                                condition="identity_type in ('X509', 'GSS', 'USERPASS')")
+                                condition="identity_type in ('X509', 'GSS', 'USERPASS', 'SAML')")
```
in the downgrade, i think you ought to remove SAML here and keep SSH :-)

(same below)

Ruturaj123 replied:
Remove SAML means delete SAML from account map and identities table or just remove SAML in the condition above? As of now, SSH keys are deleted from the table so instead of that SAML credentials need to be deleted?

mlassnig replied:
the downgrade should bring the database in the same state as it was before. what you are effectively doing here is adding the SAML type in the upgrade, then when you downgrade, instead of removing the SAML you removed the SSH.

mlassnig replied:
this applies to the whole migrate script

Ruturaj123 replied:
Oh okay, I got it. Sorry Sir.

bari12 replied:
Please create a new migrate script for the adding of `SAML`. This should not be appended to the existing one 😄

Ruturaj123 replied:
Sir I was trying to create a new alembic migrate script for SAML now, but it is throwing an error. After talking to hahahannes, it seems that there is a problem with the alembic version on the next branch. So, for now, I have kept this on hold and I will create a new script as soon as the bug on next is fixed. If there is any other workaround then just suggest me and I will do that asap.

hahahannes replied:
I opened #2872 for this

mlassnig commented on lib/rucio/db/sqla/util.py:
```diff
@@ -128,6 +130,8 @@ def create_root_account():
         gss_email = config_get('bootstrap', 'gss_email')
         ssh_id = config_get('bootstrap', 'ssh_identity')
         ssh_email = config_get('bootstrap', 'ssh_email')
+        saml_id = config_get('bootstrap', 'saml_id')
```
these should be added to the template in etc/

mlassnig commented on lib/rucio/web/rest/webpy/v1/authentication.py:
```diff
@@ -441,6 +450,103 @@ def GET(self):
+        :param Rucio-Username: Username as a string.
+        :param Rucio-Password: Password as a string.
+        :param Rucio-AppID: Application identifier as a string.
+        :returns: "Rucio-Auth-Token" as a variable-length string header.
```
returns X-Rucio-SAML-Auth-URL, no?

mlassnig commented on lib/rucio/web/rest/webpy/v1/authentication.py:
```diff
@@ -441,6 +450,103 @@ def GET(self):
+            header('X-Rucio-Auth-Token-Expires', date_to_str(result.expired_at))
+            return str()
+
+        SAML_PATH = '/opt/rucio/lib/rucio/web/ui/common/saml/'
```
ok, we can't really depend on a path like this. for what exactly is it needed?

Ruturaj123 replied:
This is needed as a pointer to SAML config file (settings.json) and the SP certificate & key.

Ruturaj123 replied:
Also, should we change this path or keep it as it is?

bari12 replied:
Let's load the path from the rucio.cfg where you specify it in a `[saml]` section.
Thus: `config_get('saml', 'config')` (or similar)

mlassnig commented on lib/rucio/web/ui/common/saml/advanced_settings.json:
```diff
@@ -0,0 +1,33 @@
+{
```
this will have to be customised for the CERN SAML provider, right?

Ruturaj123 replied:
Yes, it needs to be customized as per CERN SAML provider

bari12 replied:
Should we rename this specific one then to template?

mlassnig commented on lib/rucio/web/ui/common/utils.py:
```diff
@@ -20,6 +20,51 @@
+    :param data: GET or POST data
+    """
+    return {
+        'https': 'on' if request['wsgi.url_scheme'] == 'https' else 'off',
```
i'm strongly in favour of failing hard if https is not available.

Ruturaj123 replied:
So if https is not available, an error should be raised?

mlassnig replied:
ah wait, this is for everything, or only for the authentication? authentication must be encrypted, the rest doesn't matter

Ruturaj123 replied:
This function is used to create the request format for SAML login. So this is used to login to IdP

mlassnig commented on lib/rucio/web/ui/common/utils.py:
```diff
@@ -20,6 +20,51 @@
+def set_cookies(token, cookie_accounts, attribs, ui_account=None):
+    # if there was no valid session token write the new token to a cookie.
+    if token:
+        setcookie('x-rucio-auth-token', value=token, path='/')
```
cookies are case-sensitive, and for consistency we should keep them in line with the headers

mlassnig commented on lib/rucio/web/ui/common/utils.py:
```diff
@@ -20,6 +20,51 @@
+    # if there was no valid session token write the new token to a cookie.
+    if token:
+        setcookie('x-rucio-auth-token', value=token, path='/')
+        setcookie('rucio-auth-token-created-at', value=int(time()), path='/')
```
int is enough? or better long?

mlassnig commented on tools/pip-requires-client:
```diff
@@ -14,3 +14,4 @@ futures>=3.2.0; python_version < '3.0'            # Clean single-source support
 six>=1.12.0<1.13.0                                # Python 2 and 3 compatibility utilities
 boto3>=1.9.130,<1.10.0                            # S3 boto protocol (new version)
 pysftp>=0.2.9,<0.3                                # Used by sftp protocol
+#python3-saml>=1.6.0								  # Used for making SAML requests using python
```
this should go to all requires files, no?

mlassnig reviewed (commented):
looks good from an auth workflow perspective to me. still unclear how we can register the app to the CERN SAML provider, but that's for us to figure out. some tiny comments inside.

Ruturaj123 committed:
WebUI: Add permission

WebUI: Resolve syntax errors

WebUI: Syntax fix

Ruturaj123 committed:
WebUI: Add dependencies to Travis Dockerfile

WebUI: Fix Travis errors

Ruturaj123 committed:
WebUI: Change paramter in SAML tests

Ruturaj123 committed:
WebUI: Apply patch b8817427249c0657151e9ccc09d9c953a2bc02b0

WebUI: Fix test cases

WebUI: Syntax fix

WebUI: Syntax fix

WebUI: Travis fix

WebUI: Resolve Conflicts

WebUI: Resolve Conflicts

bari12 commented:
Hi @Ruturaj123 
Can you please fix the syntax and python3 tests which still fail. (Have a look on travis) Otherwise I think we can merge.

Ruturaj123 commented:
Actually Sir the syntax tests are failing because there is a problem while setting up the docker image. The package xmlsec1 is not getting installed even though I have added it in `etc/docker/travis/Dockerfile`. How can we resolve this? When I install these packages in the dev environment they are installed successfully.

bari12 commented:
Ah I see. The syntax tests are actually not executed out of the docker container, thus adding the packages there doesn't do anything. Please check with @hahahannes how to add it to the workflow of the syntax tests. I think this can be just added to one of the test prepare scripts.

hahahannes commented:
Mh but if the package is added to the travis dockerfile, pylint should also see it

Ruturaj123 commented:
So @hahahannes how can the pylint error be fixed?

hahahannes commented:
Just realized that I mixed up the docker images. So after a discussion with @tbeerman, we think the easiest would be to create an own tools/pip-requires-syntax to only install sphinx, pylint and flake and not all the dependencies for running the rucio tests. Then this error would not occur. Or we just install the required packages in the tools/install_script.sh before starting pip. Or we start the syntax tests in an own container like the rucio tests. 
But an own pip-requires-syntax would be the cleanest solution

Ruturaj123 commented:
So should I remove the packages from travis Dockerfile and add it to tools/travis/install_script.sh?

tbeerman commented:
No, leave it in the Dockerfile. If you remove it there all the DB tests would fail. Instead, also install those packages in the Travis VM. I don't know if it works but an `apt install` should work since it's running on Ubuntu. If if doesn't work a separate pip-requires-syntax, as proposed by @hahahannes, would be the way to go.

Ruturaj123 committed:
WebUI: Add dependencies to travis install script

WebUI: Persmission fix

WebUI: Persmission fix

WebUI: Syntax fix

WebUI: Input fix

WebUI: alembic script fix

WebUI: Syntax fix

WebUI: Syntax fix

tbeerman commented:
The tests failed again. You'll have to use sudo with apt-get.

tbeerman commented:
Failed again. This time it couldn't find the packages. Seems like somebody had the problem already [1]. Try a `sudo apt-get update` before the install.

[1] https://github.com/travis-ci/travis-ci/issues/7351

Ruturaj123 commented:
Yes Sir, I was going to do the same. So should I add it to before_script.sh?

tbeerman commented:
No, in the install_script. Just before you run `apt-get install`

Ruturaj123 commented:
Okay so the syntax tests are now passing. The error is in python3 tests:
```
PYLINT FAILED
cat: pylint.out: No such file or directory
```

hahahannes commented:
The pylint output is saying, that you are using the `input` statement which seems to be not compatible with python2 and 3. Try maybe:
```
import six
six.moves.input()
```

Ruturaj123 commented:
@bari12 Sir, can you please restart the travis build? I have made the changes and added a new alembic script too.

hahahannes commented:
You can login on Travis with your Github credentials. Then you can restart the tests :)

Ruturaj123 commented:
@hahahannes Sir, thank you for the help :smile:  And sorry for the trouble

Ruturaj123 committed:
WebUI: Resolve conflicts

bari12 commented on lib/rucio/api/authentication.py:
```diff
@@ -141,6 +142,26 @@ def get_ssh_challenge_token(account, appid, ip=None):
+    if not permission.has_permission(issuer=account, action='get_auth_token_saml', kwargs=kwargs):
+        raise exception.AccessDenied('User with identity %s can not log to account %s' % (saml_nameid, account))
+
+    return authentication.get_auth_token_saml(account, saml_nameid, appid, ip)
```
Change account to `InternalAccount`
Similar to line 141 in this file.

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/f1b14a8c2ac1_postgres_use_check_constraints.py:
```diff
@@ -45,7 +46,7 @@ def upgrade():
     types_to_drop = set()
     for table, column, constraint_name, constraint in (
-            ('account_map', 'identity_type', "ACCOUNT_MAP_ID_TYPE_CHK", ('X509', 'GSS', 'USERPASS', 'SSH')),
+            ('account_map', 'identity_type', "ACCOUNT_MAP_ID_TYPE_CHK", ('X509', 'GSS', 'USERPASS', 'SSH', 'SAML')),
```
I don't think this should be added there, since it is an historic migrate script. All your changes should come from the `9a1b149a2044_add_saml_identity_type.py ` migrate script you wrote.

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/f1b14a8c2ac1_postgres_use_check_constraints.py:
```diff
@@ -63,7 +64,7 @@ def upgrade():
             ('dids', 'did_type', "DIDS_TYPE_CHK", did_types),
             ('dids', 'availability', "DIDS_AVAILABILITY_CHK", ('A', 'D', 'L')),
-            ('identities', 'identity_type', "IDENTITIES_TYPE_CHK", ('X509', 'GSS', 'USERPASS', 'SSH')),
+            ('identities', 'identity_type', "IDENTITIES_TYPE_CHK", ('X509', 'GSS', 'USERPASS', 'SSH', 'SAML')),
```
See above

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/9a1b149a2044_add_saml_identity_type.py:
```diff
@@ -0,0 +1,80 @@
+    '''
+
+    if context.get_context().dialect.name in ['oracle', 'mysql']:
+        execute("DELETE FROM account_map WHERE identity_type='SAML'")  # pylint: disable=no-member
```
Please remove the `DELETE FROM` lines. Of course the downgrade wouldn't work without them, but we do all the migrate scripts with the assumptions that the admins clean the tables up themselves.

Ruturaj123 replied:
After removing this, the test cases aren't passing as downgrade is failing.

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/9a1b149a2044_add_saml_identity_type.py:
```diff
@@ -0,0 +1,80 @@
+
+    elif context.get_context().dialect.name == 'postgresql':
+        schema = context.get_context().version_table_schema + '.' if context.get_context().version_table_schema else ''
+        execute("DELETE FROM " + schema + "account_map WHERE identity_type='SAML'")  # pylint: disable=no-member
```
Same as above

bari12 commented on lib/rucio/db/sqla/migrate_repo/versions/9a1b149a2044_add_saml_identity_type.py:
```diff
@@ -0,0 +1,80 @@
+        execute("DELETE FROM " + schema + "identities WHERE identity_type='SAML'")  # pylint: disable=no-member
+
+        drop_constraint('ACCOUNT_MAP_ID_TYPE_FK', 'account_map', type_='foreignkey')
+        op.execute('ALTER TABLE ' + schema + 'identities DROP CONSTRAINT IF EXISTS "IDENTITIES_TYPE_CHK", ALTER COLUMN identity_type TYPE VARCHAR')  # pylint: disable=no-member
```
You should be able to do these just with `drop_constraint` too?

bari12 commented on tools/pip-requires:
```diff
@@ -21,3 +21,4 @@ Flask==1.0.2                                      # Python web framework
 fts3-rest-API==3.7.1; python_version == '2.7'     # FTS rest API
 M2Crypto<=0.35.2                                  # Dependency of FTS rest API; Temporary fix since 0.33 does not install on CC7
 MyProxyClient==2.1.0                              # myproxy client bindings
+#python3-saml>=1.6.0						  	  # Used for making SAML requests using python
```
Please remove

bari12 commented on tools/pip-requires-client:
```diff
@@ -14,3 +14,4 @@ futures>=3.2.0; python_version < '3.0'            # Clean single-source support
 six>=1.12.0<1.13.0                                # Python 2 and 3 compatibility utilities
 boto3>=1.9.130,<1.10.0                            # S3 boto protocol (new version)
 pysftp>=0.2.9,<0.3                                # Used by sftp protocol
+#python3-saml>=1.6.0						  	  # Used for making SAML requests using python
```
Please remove

bari12 commented on tools/pip-requires-test:
```diff
@@ -19,3 +19,4 @@ subprocess32==3.5.3; python_version <= '3.0'  # A backport of the subprocess mod
 pycodestyle==2.3.1                            # New package replacing pep8
 mock==2.0.0                                   # Mock package needed for readthedocs build for mocking pycurls
 s3cmd==2.0.2                                  # Used for testing of S3
+python3-saml>=1.6.0						  	  # Used for making SAML requests using python
```
Please remove it here and put it directly to the `setup_rucio.py`
There is the possiblity to define extras (Similar to the oracle, kerberos, etc. extras) Please specify one for `saml` with this dependency.

Ruturaj123 replied:
Removing this causes the travis syntax test case to fail as it cannot import the module.

bari12 replied:
Hi Ruturaj, yes you have to add this to the `setup.py` as an extra and then also to the Dockerfile https://github.com/rucio/rucio/blob/master/etc/docker/travis/Dockerfile Thus in the dockerfile you also install it with the extra (`pip install .[oracle,...,saml]`)

bari12 reviewed (changes requested):
Hi Ruturaj, 
Apologies for the late review comments. I just have a few minor changes, @tbeerman had already a look on the webui things and approved it. It would be awesome if you could still have a look and make the adjustments! Thanks!

Ruturaj123 commented:
Yes Sir, I will update and push the changes soon!

Ruturaj123 committed:
WebUI: Incorporare suggested changes

WebUI: Syntax fix

Ruturaj123 commented:
Also the other cases are failing with the error `Error: No such container: rucio`. What could be the reason for this?

Ruturaj123 committed:
WebUI: Update travis Dockerfile

WebUI: Update travis install script

WebUI: Make necessary changes

WebUI: Merge alembic heads

WebUI: Update alembic head

bari12 commented on lib/rucio/db/sqla/util.py:
```diff
@@ -182,13 +187,17 @@ def create_root_account():
     iaa4 = models.IdentityAccountAssociation(identity=identity4.identity, identity_type=identity4.identity_type, account=account.account, is_default=True)
 
+    # SAML authentication
+    identity5 = models.Identity(identity=saml_id, identity_type=IdentityType.SAML, email=saml_email)
```
I think we have to remove these, and you have to add them to the unit-tests.
Currently the downgrade test is not running through, because we add the new SAML identities at setup and then obviously the downgrade run is failing. Can you please change that. Otherwise it seems good to me!

ericvaandering commented:
This PR is causing a problem at this line: https://github.com/rucio/rucio/blob/870d36024db27f709d66e9b7c12974bf0edc6fe2/lib/rucio/web/rest/webpy/v1/authentication.py#L49

It assumes a "client" section exists in the config file for the server. And the server images are defined such that there is no client section.

bari12 commented:
Yes I think this is done wrong and should probably be done similar to https://github.com/rucio/rucio/blob/7e32dee2c3af1fd51417481483449ca3d9c3bf28/lib/rucio/client/baseclient.py#L67
Thus check if the extra modules are on path and only load them if they are.
@Ruturaj123 do you possibly have time to submit a PR for this minor change?

bari12 commented:
See #3209
